lib/ruby_lsp/test_unit_test_runner.rb

```diff
@@ -0,0 +1,98 @@
+# typed: true
+# frozen_string_literal: true
+
+require "test/unit"
+require "test/unit/ui/testrunner"
+require "ruby_lsp/test_reporter"
+require "ruby_indexer/lib/ruby_indexer/uri"
+
+module RubyLsp
+  class TestRunner < ::Test::Unit::UI::TestRunner
+    #: (::Test::Unit::TestCase test) -> void
+    def test_started(test)
+      current_test = test
+      @current_uri = uri_for_test(current_test)
+      return unless @current_uri
+
+      @current_test_id = "#{current_test.class.name}##{current_test.method_name}"
+      TestReporter.start_test(
+        id: @current_test_id,
+        uri: @current_uri,
+      )
+    end
+
+    #: (::Test::Unit::TestCase test) -> void
+    def test_finished(test)
+      if test.passed?
+        # tests with an Omission are still marked as passed, which seems strange
+        # return if test.instance_variable_get("@_result").faults.any?
+
+        TestReporter.record_pass(
+          id: @current_test_id,
+          uri: @current_uri,
+        )
+      end
+    end
+
+    #: (::Test::Unit::Failure | ::Test::Unit::Error | ::Test::Unit::Pending result) -> void
+    def result_fault(result)
+      case result
+      when ::Test::Unit::Failure
+        record_failure(result)
+      when ::Test::Unit::Error
+        record_error(result)
+      when ::Test::Unit::Pending
+        record_skip(result)
+      end
+    end
+
+    #: (::Test::Unit::Failure failure) -> void
+    def record_failure(failure)
+      TestReporter.record_fail(
+        id: @current_test_id,
+        type: failure.class.name,
+        message: failure.message,
+        uri: @current_uri,
+      )
+    end
+
+    #: (::Test::Unit::Error error) -> void
+    def record_error(error)
+      TestReporter.record_error(
+        id: @current_test_id,
+        message: error.message,
+        uri: @current_uri,
+      )
+    end
+
+    #: (::Test::Unit::Pending pending) -> void
+    def record_skip(pending)
+      TestReporter.record_skip(
+        id: @current_test_id,
+        message: pending.message,
+        uri: @current_uri,
+      )
+    end
+
+    #: (::Test::Unit::TestCase test) -> URI::Generic?
+    def uri_for_test(test)
+      location = test.method(test.method_name).source_location
+      return unless location # TODO: when might this be nil?
+
+      file, _line = location
+      return if file.start_with?("(eval at ") # test is dynamically defined (TODO: better way to check?)
+
+      absolute_path = File.expand_path(file, __dir__)
```

Comment:
Is this file expansion needed and correct? I would assume the return of `source_location` to already be an absolute path.

Also, we're expanding the path based on `__dir__`, which refers to the `lib/ruby_lsp` directory and is bound to result in weird paths when running tests for applications.